synapse/storage/engines/sqlite.py

```diff
@@ -88,6 +88,10 @@ def on_new_connection(self, db_conn: "LoggingDatabaseConnection") -> None:
 
         db_conn.create_function("rank", 1, _rank)
         db_conn.execute("PRAGMA foreign_keys = ON;")
+
+        # Enable WAL.
+        # see https://www.sqlite.org/wal.html
+        db_conn.execute("PRAGMA journal_mode = WAL;")
```

Comment:
is it correct that we do this on every new connection? According to https://www.sqlite.org/wal.html#persistence_of_wal_mode it seems like it is persisted in the database, so shouldn't it just go in a database schema migration file?

Reply:
Indeed, it's sufficient to do this once (although doing it every time won't cause issues). Any pointers on how to implement this as a migration?

Reply:
Have a look at https://matrix-org.github.io/synapse/develop/development/database_schema.html#delta-files.

Reply:
It doesn't look like the pragma will be dumped when we [do a full schema dump ](https://github.com/matrix-org/synapse/pull/13897#discussion_r981252643):

```
$ sqlite3 test.db
-- Loading resources from /home/dmr/.sqliterc
SQLite version 3.36.0 2021-06-18 18:36:39
Enter ".help" for usage hints.
sqlite> PRAGMA journal_mode;
journal_mode
delete
sqlite> PRAGMA journal_mode = WAL;
journal_mode
wal
sqlite> .schema
sqlite> .dump --data-only --nosys
sqlite> .dump 
PRAGMA foreign_keys=OFF;
BEGIN TRANSACTION;
COMMIT;
sqlite> 
```

So: if we effect this change via a migration, new installations won't benefit from this the next time we take a schema dump.